### PR TITLE
tests: replace jest with mocha

### LINKED ISSUE #14047: Migrate from Jest to Mocha
# Motivation

Our usage of Jest has been identified as a blocker for [migrating to ES modules](https://github.com/GoogleChrome/lighthouse/issues/12689). In short, there is a bug in V8 which we don't have a way to resolve, and even if we did, only the latest Node would be usable for our unit tests making migration impossible for some years. For more on that bug, see [this](https://github.com/facebook/jest/issues/11438#issuecomment-923835189) and [this](https://bugs.chromium.org/p/v8/issues/detail?id=10284).

Besides the ES modules story, we also wish to migrate away from Jest for other reasons–a common refrain is that tests written in Jest are verifiying that things work in a "Node + whatever Jest adds" framework, not just "Node". This may just be an act of idealism, as I don't recall running into anything significant differences in my `esm` migration branches that could be framed as "Jest not really testing Node", but it's one of those things you don't know until you explore it. At least without Jest, we can be more confident our tests aren't relying on an undocumented or not-well-understood artifact of its runtime.

# New test runner

Mocha is a barebones test runner, so we'll need to add a few things on top to make a transition from Jest feasible. Capabilities Jest offers that we need to keep:

## Expect

This is simple: although made by Jest, `expect` is a standalone npm package.

## Snapshots

Jest implements snapshots with `jest-snapshot`, which is a standalone package. We can continue to utilize the same
snapshot tests and expectation files by utilizing this package in Mocha's setup module–adding the necessary extensions via `expect.extend` and calling the update logic in a root `afterAll` lifecycle function.

## Mocks

In Jest, the term "mock" is a bit overloaded. The `jest.fn(), jest.spyOn()` is the mocking interface and acts _on objects_, and allows for expectations to be made on if/how often it was called. This is provided by `jest-mock`, and again: we can just use this package directly.

On the other hand, `jest.mock` acts _on modules_ (give it a module path and Jest's runtime will replace it with the provided mock). To replace this functionality, we can use `testdouble` (`jest.mock` -> `td.replace`). It uses just Node's `--loader` API as opposed to Jest using V8's `vm`, which is where the fatal bug mentioned above lives.

Changing what mocking library we use (or even, removing mocks entirely) has not been evaluated, and is out of scope for this migration.

## Timers

We make heavy use of Jest's fake timers... which again, is a standalone package we can just pull in: `@jest/fake-timers`.

## Test isolation

By default, Mocha runs all tests in the same process. If given the flag `--parallel`, Mocha will spin up some worker processes which run a subset of each test (note: each process will still only run a single test at a time). Because there is no v8-level module isolation, we must ensure that tests don't impact the global state (or clean up when they do).

Unfortunately, pretty much every test that mocks modules, or uses fakes times, does some setup in the root test file scope (as opposed to a root test file `beforeAll`). Mocha processes will first import/require all the test files... and then start running each test file's suite one at a time. This means that the first thing that happens in Mocha is all of our root-level mocks / fake timer stuff is applied, greatly impacting the other tests that run in that process.

A simple approch to guarentee test isolation would be to run Mocha with one test at a time, and optionally manage a worker pool to approximate the same benefit of `--parallel`. However, we may have to pay noticeable overhead with Mocha and our setup file being repeated so much. Instead, we can run all the "non-mocks, non-timers" tests together in the same call to Mocha, and isolate just the problematic tests in separate calls to Mocha.

As a followup to this work, we can migrate each of these tests to do their setup in `beforeAll`. This is being deferred because there's a chance these changes could be subtle breakages, so best to do them in their own PR.

Should note: all the `testdouble` and fake timers state can be cleared after each test suite in a [`beforeAll root hook`](https://mochajs.org/#root-hook-plugins), defined in the Mocha setup file.

## Similar CLI interface

With Jest's CLI, you can run a subset of test files by passing an argument that filters which tests files to run via a string-includes check: `jest audits` will run all the test files with "audits" in the name. To do the same with Mocha you need to use a glob: `mocha 'lighthouse-core/**/audits/*'`.

In Jest, to run test cases whose title matches a pattern, you do `yarn jest -t pattern`. With mocha, that is `mocha --grep pattern`.

These differences are enough to encourage a small wrapper around `mocha` to offer the same interface.

_____

TLDR; Let's move away from Jest by using only _parts_ of Jest. lol

# Tasks

The bulk of this work will be done in one big PR (because keeping Jest and Mocha as test runners at the same time is a headache best avoided). However, there are some smaller bits of work that can be done first.

- [x] Reach consensus on replacing Jest with Mocha
- [x] #14020 Move away from non-standard Jest imports of json files
- [x] #14036
- [ ] #14054 Migrate all Jest tests to Mocha, using #14035 as a base

A number of draft PRs have been put up.

#13568 - This was a first pass, that converts just `lighthouse-cli` tests and attempts to keep Jest and Mocha around at the same time. Don't spend very long look at this, instead...
#14018 - This is an example PR of one of the simpler test folders to migrate
#14035 - This converts `lighthouse-core` tests, expanding on the mocha setup/wrapper script from the previous attempt, and does not try to keep Jest tests working. Areas of focus here are `mocha-setup.cjs` and `run-tests.sh`

### PR DESCRIPTION
**Pull request by connorjclark**
Fixes #14047

### CODE REVIEW

File: docs/recipes/integration-test/example-lh-auth.test.js [hunk 1]
```diff
@@ -7,23 +7,22 @@
...
- * @fileoverview Example Jest tests for demonstrating how to run Lighthouse on an authenticated
+ * @fileoverview Example Mocha tests for demonstrating how to run Lighthouse on an authenticated
  * page as integration tests. See docs/recipes/auth/README.md for more.
  */
 
 /** @typedef {import('lighthouse/types/lhr')} LH */
 
 const puppeteer = require('puppeteer');
 const lighthouse = require('lighthouse');
+const {expect} = require('expect');
```
Comment (adamraine) [inline docs/recipes/integration-test/example-lh-auth.test.js:18]: According to `docs/recipes/integration-test/.eslintrc.cjs` this should be a global right?

File: flow-report/test/setup/env-setup.ts [hunk 2]
```diff
@@ -4,29 +4,48 @@
...
-    addListener: jest.fn(),
-  }));
-});
+// @ts-ignore: File is not under 'rootDir'
+import {LH_ROOT} from '../../../root.js';
+
+// https://github.com/testing-library/preact-testing-library/issues/36#issuecomment-1136484478
+fs.writeFileSync(`${LH_ROOT}/node_modules/@testing-library/preact/dist/esm/package.json`,
+  '{"type": "module"}');
+fs.writeFileSync(`${LH_ROOT}/node_modules/@testing-library/preact-hooks/src/package.json`,
```
Comment (adamraine) [inline flow-report/test/setup/env-setup.ts:20]: Looks like this was fixed https://github.com/testing-library/preact-testing-library/pull/58

File: flow-report/test/wrappers/markdown-test.tsx [hunk 3]
```diff
@@ -7,6 +7,7 @@
 import {render} from '@testing-library/preact';
 
 import {Markdown} from '../../src/wrappers/markdown';
+import '../setup/env-setup';
```
Comment (adamraine) [inline flow-report/test/wrappers/markdown-test.tsx:10]: This is included via command args right? Can we remove?

File: lighthouse-core/test/fraggle-rock/scenarios/api-test-pptr.js [hunk 4]
```diff
@@ -4,21 +4,22 @@
...
+import jestMock from 'jest-mock';
 
 import * as lighthouse from '../../../fraggle-rock/api.js';
 import {createTestState, getAuditsBreakdown} from './pptr-test-utils.js';
 import {LH_ROOT} from '../../../../root.js';
 
-jest.setTimeout(90_000);
+describe('Fraggle Rock API', function() {
+  // eslint-disable-next-line no-invalid-this
+  this.timeout(120_000);
```
Comment (adamraine) [inline lighthouse-core/test/fraggle-rock/scenarios/api-test-pptr.js:15]: Why is the test declared like this, with the named function and timeout on `this`. It doesn't match what was done in other files.

File: lighthouse-core/test/fraggle-rock/user-flow-test.js [hunk 5]
```diff
@@ -337,6 +337,7 @@ describe('UserFlow', () => {
 
       const flowResult = await auditGatherSteps(gatherSteps, {config: flowConfig});
 
+      // @ts-expect-error toMatchObject is giving a loud annoying error...
```
Comment (adamraine) [inline lighthouse-core/test/fraggle-rock/user-flow-test.js:340]: As in a TS error? Could you elaborate in this comment?

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 6]
```diff
@@ -31,16 +33,8 @@ const unresolvedPerfLog = readJson('./../fixtures/unresolved-perflog.json', impo
...
- * @param {string} moduleName
- * @return {any}
- */
-const requireMockAny = (moduleName) => {
-  return jest.requireMock(moduleName);
-};
-
-makeMocksForGatherRunner();
+/** @type {jestMock.SpyInstance<Promise<void>, [session: any, pageUrl: string]>} */
+let assertNoSameOriginServiceWorkerClientsMock;
```
Comment (adamraine) [inline lighthouse-core/test/gather/gather-runner-test.js:37]: I live for the day when we kill this file.

File: lighthouse-core/test/gather/gatherers/full-page-screenshot-test.js [hunk 7]
```diff
@@ -34,8 +32,6 @@ let screenSize;
 let screenshotData;
 let mockContext = createMockContext();
 
-jest.setTimeout(10_000);
```
Comment (adamraine) [inline lighthouse-core/test/gather/gatherers/full-page-screenshot-test.js:37]: Was this intentionally removed?

File: lighthouse-core/test/runner-test.js [hunk 8]
```diff
@@ -687,7 +688,7 @@ describe('Runner', () => {
       const AuditClass = require(auditPath);
       assert.strictEqual(AuditClass.meta.id, auditExpectedName);
     });
-  });
+  }).timeout(20_000);
```
Comment (adamraine) [inline lighthouse-core/test/runner-test.js:691]: New timeout?

File: flow-report/test/common-test.tsx [hunk 9]
```diff
@@ -4,18 +4,20 @@
  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
  */
 
-import {jest} from '@jest/globals';
+import jestMock from 'jest-mock';
 import {act, render} from '@testing-library/preact';
 
+// @ts-ignore: File is not under 'rootDir'
+import {timers} from '../../lighthouse-core/test/test-env/fake-timers.js';
```
Comment (adamraine) [inline flow-report/test/common-test.tsx:11]: Ideally this could go in `shared` or something, but that already has a `test` folder and this is the only place we do this so it's probably fine.

**Review by adamraine — COMMENTED** [review]
First pass comment dump.

There are some timeouts removed in the unit tests, was that intentional? Seems fine without timeouts but just checking.

File: flow-report/test/setup/env-setup.ts (hunk 2 above)
Comment (connorjclark) [reply flow-report/test/setup/env-setup.ts:20]: It's a similar (same?) issue, but it won't work without (at least, the preact-hooks one? maybe both)

File: flow-report/test/common-test.tsx (hunk 9 above)
Comment (connorjclark) [reply flow-report/test/common-test.tsx:11]: Moving common tests folder to a common folder will be done sometime later.

File: lighthouse-core/test/fraggle-rock/scenarios/api-test-pptr.js (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/test/fraggle-rock/scenarios/api-test-pptr.js:15]: This method sets the timeout for all nested tests, adding `.timeout(...)` to the end of the describe block seems to only effect describe / the lifecycle methods.

**Comment by connorjclark** [comment]
This coverage delta doesn't make any sense (run-test.js), maybe upgrading c8 will help

image

File: flow-report/test/run-flow-report-tests.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+ARGS=(
+  --testMatch='{flow-report/**/*-test.ts,flow-report/**/*-test.tsx}'
+  --require=flow-report/test/setup/env-setup.ts
+  --loader=@esbuild-kit/esm-loader
+  # util-test.tsx won't finish on its own because of an open MessagePort, so help it out.
+  # See https://github.com/jsdom/jsdom/issues/2448#issuecomment-802288244
+  --exit
+)
+
+yarn mocha ${ARGS[*]} "$@"
```
Comment (adamraine) [inline flow-report/test/run-flow-report-tests.sh:20]: Would be nice if we could run any test from a single command. WDYT about conditionally adding these flags in `run-tests.js`.

File: lighthouse-core/test/lib/asset-saver-test.js [hunk 11]
```diff
@@ -31,7 +31,10 @@ function assertTraceEventsEqual(traceEventsA, traceEventsB) {
 }
 describe('asset-saver helper', () => {
   describe('saves files', function() {
-    beforeAll(() => {
+    const tmpDir = `${LH_ROOT}/.tmp/asset-saver-test`;
+    fs.mkdirSync(tmpDir, {recursive: true});
```
Comment (adamraine) [inline lighthouse-core/test/lib/asset-saver-test.js:35]: Why can't this go in `before`?

File: flow-report/test/run-flow-report-tests.sh (hunk 10 above)
Comment (connorjclark) [reply flow-report/test/run-flow-report-tests.sh:20]: what do you mean? can just do `yarn unit-flow <test file>`

Comment (connorjclark) [reply flow-report/test/run-flow-report-tests.sh:20]: Oh, I see. This keeps it simpler, since only these tests need to user different loaders or requires. To do that in the other file would mean partitioning tests that need different contexts.

Comment (adamraine) [reply flow-report/test/run-flow-report-tests.sh:20]: > To do that in the other file would mean partitioning tests that need different contexts.

Yeah that's what I was suggesting. I guess it was nice when jest let you define a new test environment in `jest.config.js`.

**Review by paulirish — COMMENTED** [review]
The debugging workflow needs some love.

here's an example scenario that's very realistic for me:

* add `debugger;` within a `describe()` on lib/i18n-test.js
* run `node --inspect-brk lighthouse-core/test/scripts/run-tests.js lighthouse-core/test/lib/i18n/i18n-test.js`
* open up the node devtools window
* hit resume. 
* wait for it to hit the `debugger`…

but this doesnt work because child process.

thoughts?

File: lighthouse-core/test/scripts/run-tests.js [hunk 12]
```diff
@@ -0,0 +1,320 @@
+/**
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/test/scripts/run-tests.js:1]: i've only opened run-tests.js dozens of times now

File: .github/workflows/devtools.yml [hunk 13]
```diff
@@ -66,6 +66,8 @@ jobs:
     - name: Download DevTools Frontend
       run: bash $GITHUB_WORKSPACE/lighthouse/lighthouse-core/test/chromium-web-tests/download-devtools.sh
 
+    - run: npx patch-package # TODO: remove when mocha replaces jest
+      working-directory: ${{ github.workspace }}/lighthouse
```
Comment (adamraine) [inline .github/workflows/devtools.yml:70]: What are we patching here? Sorry if I missed it PR is big.

**Review by adamraine — APPROVED** [review]

File: package.json [hunk 14]
```diff
@@ -35,24 +35,24 @@
...
-    "unit-flow": "yarn jest \"flow-report/.*-test.[tj]s[x]?\"",
-    "unit": "yarn jest",
-    "unit:ci": "NODE_OPTIONS=--max-old-space-size=8192 npm run jest --ci .",
+    "unit-core": "yarn mocha lighthouse-core",
+    "unit-cli": "yarn mocha --testMatch 'lighthouse-cli/**/*-test.js'",
+    "unit-report": "yarn mocha --testMatch 'report/**/*-test.js'",
+    "unit-treemap": "yarn mocha --testMatch 'treemap/**/*-test.js'",
+    "unit-viewer": "yarn mocha --testMatch 'viewer/**/*-test.js'",
+    "unit-flow": "sh flow-report/test/run-flow-report-tests.sh",
+    "unit": "yarn mocha",
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply package.json:54]: Well, flags won't carry over. Maybe in a followup with a wrapper / whatever you suggest to address your other feedback?

File: lighthouse-core/test/test-env/mocha-setup.js [hunk 15]
```diff
@@ -0,0 +1,210 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/test/test-env/mocha-setup.js:2]: File was written in 2021, in general I'm not keen on updating the year esp. since there's a general sense these don't particularly matter... we just do it because we always have...

File: package.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply package.json:54]: Are the flow tests even being run in CI though? We can still have `yarn mocha` which will carry over flags, and `yarn unit` can just be the "run everything normally" command.

Or maybe add a `unit-all` for CI?

**Comment by connorjclark** [comment]
The glob / test filter is not working on Windows (0 tests running), gotta look into that.

**Comment by paulirish** [comment]
> > [The debugging workflow needs some love. […] ](https://github.com/GoogleChrome/lighthouse/pull/14054#pullrequestreview-1020898096)
> 
> re: debugging
> 
> I've had no issues with debugging, but that's because my new workflow is using vscode's JS Debugger terminal. Give that a shot. It auto attaches to child processes. Is there perhaps a flag to auto attach sub processes when calling `node` directly?

mkay i'll use the [new fancy thing instead](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_javascript-debug-terminal).

File: lighthouse-core/test/scripts/run-mocha-tests.js [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+  if (filterFilePatterns.length) {
+    console.log(`applied test filters: ${JSON.stringify(filterFilePatterns, null, 2)}`);
+  }
+  console.log(`running ${filteredTests.length} test files`);
+
+  return filteredTests;
+}
+
+const baseArgs = [
+  '--require=lighthouse-core/test/test-env/mocha-setup.js',
```
Comment (paulirish) [inline lighthouse-core/test/scripts/run-mocha-tests.js:213]: what about a [.mocharc.js file](https://github.com/mochajs/mocha/blob/master/example/config/.mocharc.js)?

is going full CLI flag just easier given the permutations?

Comment (connorjclark) [reply lighthouse-core/test/scripts/run-mocha-tests.js:213]: That file just extracts what you'd pass on the CLI to a file. It doesn't do what jest was doing, which was setting up its own environment for just for particular tests. That's the entire thing run-mocha-tests.js is doing. so in short:

> is going full CLI flag just easier given the permutations?

Yeah

**Comment by connorjclark** [comment]
Windows is surely a dumb mistake in the process execution or paths or something, but I'm wasting way too much time trying to get a windows cloudtop to work at the office. I'll disable for now, and re-enable tomorrow when I'm back home with a windows machine to debug.

#14171
